src/sinks/s3_common/sink.rs

```diff
@@ -1,68 +1,221 @@
     request_builder: RB,
-    partitioner: S3KeyPartitioner,
+    partitioner: Part,
+    transformer: Transformer,
+    serializer: Serializer,
+    framer: Framer,
     batcher_settings: BatcherSettings,
 }
 
-impl<Svc, RB> S3Sink<Svc, RB> {
+impl<Svc, RB, Part> S3Sink<Svc, RB, Part> {
     pub const fn new(
         service: Svc,
         request_builder: RB,
-        partitioner: S3KeyPartitioner,
+        partitioner: Part,
+        transformer: Transformer,
+        serializer: Serializer,
+        framer: Framer,
         batcher_settings: BatcherSettings,
     ) -> Self {
         Self {
             partitioner,
             service,
             request_builder,
+            transformer,
+            serializer,
+            framer,
             batcher_settings,
         }
     }
 }
 
-impl<Svc, RB> S3Sink<Svc, RB>
+pub struct EncodedEvent {
+    pub original: Event,
+    pub json_size: GroupedCountByteSize,
+    pub encoded: BytesMut,
+}
+
+impl Finalizable for EncodedEvent {
+    fn take_finalizers(&mut self) -> vector_common::finalization::EventFinalizers {
+        // TODO: Do we need to worry about finalizer getting cloned prior to encoding?
+        self.original.take_finalizers()
+    }
+}
+
+pub trait RequestBuilder {
+    type Request;
+    type Metadata;
+    type PartitionKey;
+
+    /// Compression to be used when building the payload
+    fn compression(&self) -> Compression;
+
+    /// Calculate any necessary request metadata from the original events
+    fn build_metadata(
+        &self,
+        partition_key: Self::PartitionKey,
+        events: Vec<Event>,
+    ) -> Self::Metadata;
+
+    /// Build a full request given metadata and payload.
+    ///
+    /// `RequestMetadata` here is mostly just a pass-thru field that we should try to factor out
+    /// into a common layer.
+    fn build_request(
+        &self,
+        metadata: Self::Metadata,
+        payload: Bytes,
+        request_metadata: RequestMetadata,
+    ) -> Self::Request;
+}
+
+/// Handles the logic for turning an event into an encoded byte string. This encapsulates both
+/// `Transformer` (to handle `{only,except}_fields`) and `Serializer` from the codecs crate, as well
+/// as the calculation of estimated JSON size that needs to happen between those two phases.
+///
+/// We use an `Encoder` around the serializer here because that's where the API is exposed right
+/// now. Ideally this struct lets us isolate the complexity of that interface to make it easier to
+/// change in the future.
+struct EventEncoder {
+    transformer: Transformer,
+    encoder: crate::codecs::Encoder<()>,
+}
+
+impl EventEncoder {
+    const fn new(transformer: Transformer, serializer: Serializer) -> Self {
+        let encoder = crate::codecs::Encoder::<()>::new(serializer);
+        Self {
+            transformer,
+            encoder,
+        }
+    }
+
+    fn encode(&mut self, mut event: Event) -> EncodedEvent {
+        let original = event.clone();
```

Comment:
It is a shame that we have to do this, but I'm not seeing any quick way around it.

If we could change the serializer so it didn't take ownership of the event, but could instead take a view of a reference to the event and that view incorporated only the fields not dropped by the transformer.. Probably a significant amount of work... if even possible.

Reply:
Yeah, my other idea (also a significant amount of work) was to somehow have things like the partitioner and request builder declare what fields they'd need later so we could save and pass specifically those rather than the whole event.

On the other hand, at least for log events, if we don't actually modify the event in the transformer (i.e. no `{only,except}_fields` config then the clone here should just be an Arc clone of the root value which is basically free. So I'm hopeful that in most cases this won't really have any performance impact at all.

Reply:
True, and the transformer will be deprecated with LogNamespacing coming in, so this isn't going to be much of an issue for long anyway.

Reply:
Could we store the transformed event, or is that not useful for what this original field is for?